### PR TITLE
Update from facebook

### PR DESCRIPTION
**Pull request by sf-wind**

### CODE REVIEW

**Review by soumith — DISMISSED** [review]
Declarations.yaml shouldn't be included. It's auto-generated.

**Comment by soumith** [comment]
commit `Skip calling ncclCommDestroy in destructor 0a6dd52` is already in pytorch master, it should be removed from this PR

**Comment by soumith** [comment]
`Fix ATenOp dispatch for ops with TensorList arg` is already in master via https://github.com/pytorch/pytorch/pull/8226 , so can be removed.

**Comment by soumith** [comment]
I believe "Resolve conflicts for tools/jit/gen_jit_dispatch.py  b8e2d3a" is already in master as well. cc: @jamesr66a

File: aten/src/ATen/Declarations.yaml [hunk 1]
```diff
@@ -22710,7 +22710,7 @@
   abstract: true
   auto_gpu: true
   with_gil: false
-- name: _cast_uint8_t
+- name: _cast_Byte
```
Comment (soumith) [inline aten/src/ATen/Declarations.yaml]: this commit looks sketchy as well. cc: @gchanan 
I'm not sure this change should've gone into this PR

File: torch/_thnn/utils.py [hunk 2]
```diff
@@ -2,8 +2,8 @@
 import itertools
 import importlib
 
-THNN_H_PATH = os.path.join(os.path.dirname(__file__), '..', 'lib', 'THNN.h')
```
Comment (soumith) [inline torch/_thnn/utils.py]: these changes will come in via https://github.com/pytorch/pytorch/pull/8214, so you can remove them from this PR as well.

File: tools/jit/gen_jit_dispatch.py [hunk 3]
```diff
@@ -100,6 +100,8 @@ def is_jit_op(decl):
     # Only support a single TensorList arg
     if sum(arg['simple_type'] == 'TensorList' for arg in arguments) > 1:
         return False
+    if any(arg['simple_type'] == 'std::string' for arg in arguments):
```
Comment (soumith) [inline tools/jit/gen_jit_dispatch.py]: these changes can be removed as well. cc: @jamesr66a

**Comment by ezyang** [comment]
```

23:49:52 [ RUN      ] NetDeathTest.DeclaredOutputNotMet
23:49:52 ./.jenkins/caffe2/test.sh: line 60:    25 Segmentation fault      (core dumped) "$test" --gtest_output=xml:"$gtest_reports_dir/$(basename $test).xml"
```

File: tools/jit/gen_jit_dispatch.py (hunk 3 above)
Comment (sf-wind) [reply tools/jit/gen_jit_dispatch.py]: This change is not yet in master. Should this be removed?

Comment (jamesr66a) [reply tools/jit/gen_jit_dispatch.py]: Remove

File: tools/autograd/utils.py [hunk 4]
```diff
@@ -8,11 +9,8 @@
     'split_name_params', 'write',
 ]
 
-try:
-    from src.ATen.code_template import CodeTemplate
-except ImportError:
-    from tools.shared.module_loader import import_module
-    CodeTemplate = import_module('code_template', 'aten/src/ATen/code_template.py').CodeTemplate
+
+CodeTemplate = import_module('code_template', 'aten/src/ATen/code_template.py').CodeTemplate
```
Comment (soumith) [inline tools/autograd/utils.py]: these changes came back while rebasing, remove them

File: tools/nnwrap/generate_wrappers.py [hunk 5]
```diff
@@ -109,12 +109,7 @@ def wrap_nn(thnn_h_path, install_dir, template_path):
     for fn in nn_functions:
         for t in ['Float', 'Double']:
             wrapper += wrap_function(fn.name, t, fn.arguments)
-    install_dir = install_dir or 'torch/csrc/nn'
-    try:
-        os.makedirs(install_dir)
-    except OSError:
-        pass
-    with open(os.path.join(install_dir, 'THNN.cwrap'), 'w') as f:
+    with open('torch/csrc/nn/THNN.cwrap', 'w') as f:
```
Comment (soumith) [inline tools/nnwrap/generate_wrappers.py]: these changes came back while rebasing, remove them

**Review by soumith — APPROVED** [review]
merge after tests pass

**Comment by Yangqing** [comment]
Did a bit more digging and it seems that tooling is fine. Basically fb internal diff landed on 3/7 afternoon (adding the c10 folder), and the github PR landed 3/7 night (strictly speaking 3/8), and because the PR added and then deleted the c10 folder during the discussion, the final landing was not aware of the c10 folder situation at all in the single squash and merge. And as a result, it was basically that a dangling c10 folder need manual deletion - sync script itself is correct.

**Pull request merged**